app/controllers/concerns/idv_session.rb

```diff
@@ -12,12 +12,37 @@ def confirm_idv_session_started
     redirect_to idv_session_url unless idv_session[:params].present?
   end
 
+  def confirm_idv_attempts_allowed
+    if idv_attempter.exceeded?
+      flash[:error] = t('idv.errors.hardfail')
+      redirect_to idv_fail_url
+    elsif idv_attempter.reset_attempts?
+      self.idv_attempts = 0
+    end
+  end
+
   def idv_session
     user_session[:idv] ||= {}
   end
 
   protected
 
+  def idv_attempts=(num)
+    current_user.update!(idv_attempts: num)
+  end
+
+  def idv_attempts
+    current_user.idv_attempts
+  end
+
+  def idv_attempter
+    @_attempter ||= IdvAttempter.new(current_user)
+  end
+
+  def idv_flag_user_attempt
+    current_user.update!(idv_attempted_at: Time.zone.now)
+  end
+
```

Comment:
Yes totally agreed. It has outgrown it's original scope. Let's do it in a separate PR so we don't conflate the changes though.